core/src/test/scala/unit/kafka/log/LogValidatorTest.scala

```diff
@@ -147,6 +148,43 @@ class LogValidatorTest {
       compressed = true)
   }
 
+  @Test
+  def testInvalidOffsetRangeAndRecordCount(): Unit = {
+    // The batch to be written contains 3 records, so the correct lastOffsetDelta is 2
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 0, count = 3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 15, count = 3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = -3, count = 3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 2, count = -3)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 2, count = 6)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 2, count = 0)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = -3, count = -2)
+
+    // Count and offset range are consistent with each other, but do not match the actual number of records
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 5, count = 6)
+    validateRecordBatchWithCountOverrides(lastOffsetDelta = 1, count = 2)
```

Comment:
Do we also want a test here with `lastOffsetDelta = 2, count = 3` that doesn't throw an exception (for completion even though it is tested elsewhere)?